nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,169 @@
+#  Copyright 2019 SURF.
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#          http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+import hashlib
+import hmac
+import pickle
+from functools import wraps
+from typing import Any, Callable
+
+import structlog
+from redis.asyncio import Redis as AIORedis
+
+logger = structlog.get_logger(__name__)
+
+
+def get_hmac_checksum(secret, message):
+    # Todo: check if the current secret is good enough for sha512, it might be better to use a new SECRET just for this
+    h = hmac.new(secret.encode(), message, hashlib.sha512)
+    return h.hexdigest()
+
+
+async def set_signed_cache_value(pool: AIORedis, secret: str, cache_key: str, value: Any, expiry: int):
+    pickled_value = pickle.dumps(value)
+    checksum = get_hmac_checksum(secret, pickled_value)
+    pipeline = pool.pipeline()
+    pipeline.setex(cache_key, expiry, pickle.dumps(value)).setex(f"{cache_key}-checksum", expiry, checksum.encode())
+    result1, result2 = await pipeline.execute()
+    if not result1 or not result2:
+        logger.warning("Cache not set", cache_key=cache_key, value_ok=result1, checksum_ok=result2)
+
+
+async def get_signed_cache_value(pool: AIORedis, secret: str, cache_key: str):
+    pipeline = pool.pipeline()
+    pipeline.get(cache_key).get(f"{cache_key}-checksum")
+    pickled_value, cache_checksum = await pipeline.execute()
+    if not pickled_value or not cache_checksum:
+        return None
+
+    cache_checksum = cache_checksum.decode("utf-8")
+    if (checksum := get_hmac_checksum(secret, pickled_value)) != cache_checksum:
+        logger.error(
+            "Checksum for cache was wrong, someone tampered with the values!",
+            correct_checksum=str(cache_checksum),
+            recalculated_checksum=checksum,
+        )
+        return None
+
+    return pickle.loads(pickled_value)
+
+
+def cached_result(pool: AIORedis, prefix: str, secret: str, key_name: str | None, expiry: int = 120) -> Callable:
+    """Pass returned result objects from a function call into redis.
+
+    Returns a decorator function that will cache every result of a function to redis. This only works
+    for functions with string, int or UUID arguments and result objects that can be serialized by the
+    python pickle library.
+
+    Example:
+        Todo: switch to cPickle?
```

Comment:
Perhaps add a small benchmark script to test the overhead on various types and sizes of cached values?
Then we can compare to cPickle (or perhaps orjson)

Reply:
cPickle is used by default, under the hood, in python 3.x
https://www.geeksforgeeks.org/serializing-data-using-the-pickle-and-cpickle-modules/

So removed the Todo: but will write a benchmark regarding JSON vs Pickle